src/hotspot/share/opto/loopopts.cpp

```diff
@@ -341,24 +341,9 @@ void PhaseIdealLoop::dominated_by(IfProjNode* prevdom, IfNode* iff, bool flip, b
   // dominating projection.
   Node* dp = iff->proj_out_or_null(pop == Op_IfTrue);
 
-  // Loop predicates may have depending checks which should not
-  // be skipped. For example, range check predicate has two checks
-  // for lower and upper bounds.
   if (dp == nullptr)
     return;
```

Comment:
Since we bail out above if `iff->outcnt() != 2` (can it even be that we have an `If` at this point which does not have 2 out projections?) this bailout seems redundant. Looks like it was only added due to a parfait report with https://github.com/openjdk/jdk/commit/25c4a7fccdbdaa9da0a7aa5e04e80966138fe42c. Maybe we can remove that as well and change `proj_out_or_null()` back to  `proj_out()` (not sure though if parfait will then report this again). But could also be done separately.

Reply:
Thanks for the details. Why would it have it been necessary before but no longer necessary now? What is it that has changed so parfait would not complain?

Reply:
Unfortunately, the report details are no longer available today. I think the fix back there should have been that it's a false positive and it cannot happen that `dp` is null - even though parfait fails to prove that (it probably still cannot). I'm not exactly sure how parfait works and how we could ensure that it will not complain about this again but maybe adding an assert that `dp` is not null would help. Anyway, this should not block this PR and might be better handled separately.